sklearn/ensemble/_base.py

```diff
@@ -151,6 +151,12 @@ def _make_estimator(self, append=True, random_state=None):
         estimator.set_params(**{p: getattr(self, p)
                                 for p in self.estimator_params})
 
+        # TODO: Remove in v1.2
+        # criterion "mse" would cause a warnings in every call to
+        # DecisionTreeRegressor.fit(..)
+        if getattr(estimator, "criterion", None) == "mse":
+            estimator.set_params(criterion="squared_error")
```

Comment:
Estimators such as `BaggingClassifier` accepts a `base_estimator` parameter. This means that third-party estimators passed as `base_estimator` could have a "criterion" parameter that does not support "squared_error". Currently, I see two ways around this:

1. The hacky workaround would be to detect the classes we made the name change to and only make this change for those classes.

2. A cleaner solution would be to remove `criterion` from `self.estimator_params` and have a "additional_estimator_params" kwarg in `_make_estimator`. This way the caller can set a `self.criterion_` and pass it into `_make_estimator`. This would be...more code tho and maybe a little over-engineered for something we are going to remove anyways.

Reply:
As we're going to remove this piece of code anyway, I'd go for solution 1. Deprecation for `criterion="mse"` was made for:
- ExtraTreesRegressor
- RandomForestRegressor
- DecisionTreeRegressor
- ExtraTreeRegressor

Reply:
The problem is that RandomForestRegressor inherits from BaseEnsemble. Suggestions?

Reply:
+1 for the first solution ("detect the classes we made the name change to and only make this change for those classes"). It feels like that a saner and more predictable way to handle deprecation.

But actually, I don't understand why we even need to do this. Why not just let the warning passthrough if the user has code with a base estimator explicitly constructed with `criterion="mse"`?

If they just use the default, they should not get any warning, no?

Reply:
When a user sets `RandomForestRegressor(criterion='mse'), this PR currently raises a warning in `RandomForestRegressor.fit`. Later in `_make_estimator`, `criterion` would be passed down to base estimators through `self.estimator_params`. This would raise warnings again when the base estimator `fit` methods are called.

> The problem is that RandomForestRegressor inherits from BaseEnsemble. Suggestions?

I think we only need to reset criterion for `RandomForestRegressor` and `ExtraTreesRegressor`, where `base_estimator` can not be passed in and criterion can be passed in. So for maintainability, we only need to detect for `ExtraTreeRegressor` and `DecisionTreeRegressor` in `_make_estimator`.

Reply:
Another "correct" OOP solution would be to remove "criterion" from `estimator_params` in `RandomForestRegressor` and `ExtraTreesRegressor`, and when we warn about `criterion` in `BaseForest.fit`, we set `criterion` correctly:

```python
self._validate_estimator()
# TODO: Remove in v1.2
if isinstance(self, (RandomForestRegressor, ExtraTreesRegressor)) self.criterion == "mse":
    warn(...)
	self.base_estimator_.criterion = "squared_error"
```

Reply:
Can you check again with https://github.com/scikit-learn/scikit-learn/pull/19310/commits/7d220b89cd2a29f9f35ea2b6ac95a200c116ed5b?